docs/decisions/0005-PII-leakage-prevention.rst

```diff
@@ -0,0 +1,59 @@
+PII leakage prevention
+================================
+
+Status
+------
+
+Pending
+
+Context
+-------
+
+Event transformers (xAPI and Caliper) fetch data from edx events that may contain PII which is not meant to be shared with learning record consumers. New event transformers are expected to be written in future and a mechanism needs to be put in place to avoid leakage of PII within these transformers.
+
+Decision
+--------
+
+#. A blocklist (one each for xAPI and Caliper) of all fields that are considered to be PII, including the ones listed in `OEP-30`_ proposal doc, will be created and maintained by edx developers and contributors (in ``settings.py``). For now, the exact name of these keys are extracted from:
+
+   #. Data References `6.1`_, `6.2`_, `6.3`_ and `6.4`_ in Edx research `guide`_.
+
+   #. JSON documents of edx events using for testing transformers in ``event-routing-backends``.
+
+#. ``strip_blocked_keys`` method in `base_transformer.py`_ will remove keys in the transformed event that are enlisted in the blocklist.
+
+#. Keys like ``name`` may or may not contain PII values. For example:
+
+   #. xAPI: ``Actor[name]`` will most likely be a PII especially if ``Actor[objectType]`` is ``Agent``.
+
+   #. xAPI: ``Object[definition][name]`` will most likely be PII if ``Object[objectType]`` is ``Agent``.
+
+   #. xAPI: ``Object[definition][name]`` will not be PII if ``Object[objectType]`` is ``Activity``.
+
+   #. Caliper: ``Actor[name]`` will most likely be PII if ``Actor[type]`` is ``Person``.
+
+   #. Caliper: ``Object[name]`` will not be PII if ``Object[type]`` is ``Page``.
+
+#. Therefore, for some keys like ``name``, we need to strip keys only if their sibling keys and/or attribute of their parent meet a specified criteria.
+
+#. Entries in each blocklist can be of type ``string`` or ``list`` with 2 elements. If the entry is ``string``, that entry (key) will be stripped off of the transformed event unconditionally.
+
+#. If the entry is type ``list``, the first element will specify the path of the key that needs to be striped, and second element will specify the condition in key value pairs. For example:
+
+   #. xAPI: If an entry is ``['Actor.name', 'Actor.objectType':'Agent']``, then the ``Actor[name]`` key will be stripped off if ``Actor[objectType]`` is ``Agent``.
```

Comment:
@ayub02 
Instead of having a list of string or list I think we can have dictionary of key and conditions. If conditions exist we should match all the conditions before string that key.
```
BLOCKLIST= [
    {'key': 'ip_address'},
    {'key': 'object.name', 'conditions': [{'key': 'object.type', 'value': 'actor'}]},
]
```